imports/server/mediasoup.ts

```diff
@@ -42,6 +49,10 @@ const mediaCodecs: types.RtpCodecCapability[] = [
   },
 ];
 
+const heartbeatFrequency = 100;
+const heartbeatTimeout = heartbeatFrequency * 10;
```

Comment:
1 second is probably a bit aggressive, given that we run stuff in the cloud where you can just get random hypervisor stolen time, where latency can be variable, and where I'm not sure what our behavior is if a server gets GC'd while still running.  We should probably consider:

* loosening the timeout a bit or
* convincing ourselves we understand what would happen if we incorrectly GC'd a server that still thinks it's alive/may be still serving clients.  Would we want some sort of STONITH/fencing mechanism if a server observes that its Servers record got deleted by another server to just intentionally exit so that it'd get restarted and take a new server ID, rather than trying to carry on in an inconsistent state?

Reply:
I'll say up front that I'm moderately opposed to introducing a STONITH mechanism, because I am always worried about the cascading failure of every node decides to shoot every other node. So I'd prefer to avoid that.

A server which has its server record deleted out from under it will actually mostly recover on its own - the server record is managed with an upsert, so it'll just get re-inserted the next time the server's periodic process runs. For any rooms that it was responsible for, it'll observe that the `Room` record was deleted out from under it and destroy the corresponding mediasoup router, which is fine - the room should have been re-assigned to a different server.

However, it won't re-insert any records tied to that `serverId` and created by active subscriptions, which currently is specifically `Peer` and `Subscriber` records, so those will disappear. Active peers will get stuck in a "no self peer" state, and active viewers will disappear from the viewer list. Not ideal.

I'm fine bumping the timeouts back up to 5s (and probably 10s for initial handshake). Either way, I would love if we could come up with some way to handle `Peer` and `Subscriber` records more gracefully, although nothing is really coming to mind at the moment.

Reply:
> I'll say up front that I'm moderately opposed to introducing a STONITH mechanism, because I am always worried about the cascading failure of every node decides to shoot every other node. So I'd prefer to avoid that.

Totally reasonable.

> However, it won't re-insert any records tied to that serverId and created by active subscriptions, which currently is specifically Peer and Subscriber records, so those will disappear. Active peers will get stuck in a "no self peer" state, and active viewers will disappear from the viewer list. Not ideal.
> 
> I'm fine bumping the timeouts back up to 5s (and probably 10s for initial handshake). Either way, I would love if we could come up with some way to handle Peer and Subscriber records more gracefully, although nothing is really coming to mind at the moment.

I suspect we'll have to involve the client, since the client will still expect terminating the subscription to trigger cleanup of their Peer or Subscriber record if we quietly fixed things up behind the scenes, so that would have to happen from the backend they're talking to.  On the bright side, for e.g. the Peer record, the client should be able to observe that their intended call state is still joined, that `joinSubRef.current` is still valid, and that they observed `selfPeer` transition from truthy to falsy.  I'm not certain exactly what the code would look like, but I suspect we can make the client notice that it got dropped and re-establish a new subscription.

With the switch to the 5s timeout/10s initial timeout, I'm no longer super concerned with this misfiring, so let's go ahead and merge this.

We normally only operate with one backend instance during the off-season, right?  So we'll probably not really see anything that goes wrong here until we scale up for 2024?

Reply:
I was hypothesizing that we could detect this in the `mediasoup:join` subscription - after creating the peer, create an observer on the peer and if it gets deleted but the subscription hasn't been stopped, just recreate it. (And do something similar in `subscribers.inc`).

I haven't decided if I think there's meaningful risk of a server being alive and thinking it still had a client connected to it, but the client has gone away - screw cases there include that the client is just gone, or that the client has reconnected to a different server and we're in the "clean up an old peer" section of the code.

That's correct that we only run with 1 backend, although I think realistically the bigger issue is that we just don't _use_ Jolly Roger in the off-season. (We could pretty affordably run with, like, 2 `t3a.micro` instead of just 1 if we would actually get good data out of it)